cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/artifact/AuthorizationArtifactRepository.java

```diff
@@ -0,0 +1,345 @@
+  }
+
+  @Override
+  public void deleteArtifactProperties(Id.Artifact artifactId) throws Exception {
+    authorizationEnforcer.enforce(artifactId.toEntityId(), authenticationContext.getPrincipal(), Action.ADMIN);
+    delegate.deleteArtifactProperties(artifactId);
+  }
+
+  @Override
+  public void addSystemArtifacts() throws Exception {
+    // to add system artifacts, users should have write privileges on the system namespace
+    Principal principal = authenticationContext.getPrincipal();
+    authorizationEnforcer.enforce(NamespaceId.SYSTEM, principal, Action.WRITE);
+    delegate.addSystemArtifacts();
+  }
+
+  @Override
+  public void deleteArtifact(Id.Artifact artifactId) throws Exception {
+    // for deleting artifacts, users need admin privileges on the artifact being deleted.
+    Principal principal = authenticationContext.getPrincipal();
+    authorizationEnforcer.enforce(artifactId.toEntityId(), principal, Action.ADMIN);
+    delegate.deleteArtifact(artifactId);
+  }
+
+  @Override
+  public List<ArtifactInfo> getArtifactsInfo(NamespaceId namespace) throws Exception {
+    List<ArtifactInfo> artifactInfos = delegate.getArtifactsInfo(namespace);
+    // todo - CDAP-11560 should filter in artifact store
+    return Collections.unmodifiableList(filterAuthorizedArtifactInfos(artifactInfos, namespace));
+  }
+
+  /**
+   * Ensures that the logged-in user has at least one {@link Action privilege} in the expected action set
+   * on the specified program instance.
+   *
+   * @param artifactId the {@link ArtifactId} to check for privileges
+   * @param actionSet the expected {@link Set} of privileges
+   * @throws UnauthorizedException if the logged in user has no {@link Action privileges} on the specified entity
+   */
+  private void ensureOnePrivilege(ArtifactId artifactId, Set<Action> actionSet) throws Exception {
+    // No authorization for system artifacts
+    if (NamespaceId.SYSTEM.equals(artifactId.getParent())) {
+      return;
+    }
+
+    boolean isAuthorized = false;
+    Principal principal = authenticationContext.getPrincipal();
+    for (Action action : actionSet) {
+      try {
+        authorizationEnforcer.enforce(artifactId, principal, action);
+        isAuthorized = true;
+        break;
+      } catch (UnauthorizedException e) {
+        // continue to next action
+      }
+    }
+    if (!isAuthorized) {
+      throw new UnauthorizedException(principal, artifactId, actionSet);
+    }
+  }
+
+  /**
+   * Filter a list of {@link ArtifactSummary} that ensures the logged-in user has a {@link Action privilege} on
+   *
+   * @param artifacts the {@link List<ArtifactSummary>} to filter with
+   * @param namespace namespace of the artifacts
+   * @return filtered list of {@link ArtifactSummary}
+   */
+  private List<ArtifactSummary> filterAuthorizedArtifacts(List<ArtifactSummary> artifacts, final NamespaceId namespace)
+    throws Exception {
+    return Lists.newArrayList(
```

Comment:
We already create a new array list here, so no need to create a new array list again when this method is called.